src/rules/at-root-no-redundant-selector/index.js

```diff
@@ -0,0 +1,74 @@
+"use strict";
+
+const { utils } = require("stylelint");
+const namespace = require("../../utils/namespace");
+const ruleUrl = require("../../utils/ruleUrl");
+
+const ruleName = namespace("at-root-no-redundant-selector");
+
+function isInterpolated(string) {
+  const interpolationRegex = /#{(.*?)}/g;
+  return interpolationRegex.test(string);
+}
+
+function isAtRootRuleNested(node) {
+  // Recursive method to check if any descendant is an @at-root rule.
+  let isAtRootNested = false;
+  if (node.nodes !== undefined) {
+    for (const i in node.nodes) {
+      isAtRootNested = isAtRootNested || isAtRootRuleNested(node.nodes[i]);
+      if (isAtRootNested) break;
+    }
+  }
+  return isAtRootNested || (node.type === "atrule" && node.name === "at-root");
+}
+
+const messages = utils.ruleMessages(ruleName, {
+  rejected: "Unnecessary @at-root rule"
+});
+
+const meta = {
+  url: ruleUrl(ruleName)
+};
+
+function rule(actual) {
+  return (root, result) => {
+    const validOptions = utils.validateOptions(result, ruleName, {
+      actual
+    });
+
+    if (!validOptions) {
+      return;
+    }
+
+    root.walkAtRules(decl => {
```

Comment:
Optional: you can pass a filter to `walkAtRules`, for example:

```js
root.walkAtRules('at-root', node => { ... })
```

Doing so, you can omit the checks for `node.name === 'at-root'` because it would be guaranteed to be this node. However, you won't be able to match `@keyframes`. But.. you can always iterate up, for example, if I want to check whether an at-root exists within a `@keyframes` node, I could do:

```js
function isWithinKeyframes(node) {
  let parent = node.parent;
  while (parent != null) {
    if (parent.type === 'at-rule' && parent.name === 'keyframes') {
      return true;
    }
    parent = node.parent;
  }
  return false;
}
```

With this, your overall implementation could be greatly reduced to:

```js
function rule(actual) {
  return (root, result) => {
    const validOptions = utils.validateOptions(result, ruleName, {
      actual
    });

    if (!validOptions) {
      return;
    }

    root.walkAtRules('at-root', node => {
      if (
        node.parent.type === 'root' ||
        node.params.replace(/#{.*}/g, '').includes('&') ||
        isWithinKeyframes(node)) {
          utils.report({
            message: messages.rejected,
            node,
            result,
            ruleName
          });
        }
    });
}
```

Reply:
Awesome!! Thanks